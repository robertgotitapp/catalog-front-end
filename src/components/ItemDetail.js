import React, { Component } from 'react';
import { connect } from 'react-redux';
import { updateItem, removeItem } from '../actions/items';

export class ItemDetail extends Component {
  render() {
    return (
      <div>
        {item.name}
      </div>
    );
  }
}

function mapStateToProps({ itemsReducer }, { match }) {
  return {
    item: itemsReducer.items[match.params.id],
  };
}

const mapDispatchToProps = {
  updateItem,
  removeItem,
};

export default connect(mapStateToProps, mapDispatchToProps)(ItemDetail);
