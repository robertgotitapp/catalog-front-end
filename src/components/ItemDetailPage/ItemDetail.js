import React, { Component } from 'react';
import { connect } from 'react-redux';
import Button from 'react-bootstrap/Button';
import { Link } from 'react-router-dom';
import Alert from 'react-bootstrap/Alert';
import { removeItem, getItems } from '../../actions/items';
import { PaginationConfig } from '../../utils/const';

export class ItemDetail extends Component {
  state = {
    alerts: '',
  }

  deleteItem = (e) => {
    e.preventDefault();
    this.props.removeItem(
      this.props.item.category_id,
      this.props.item.id,
    )
      .then((res) => {
        if (res.statusCode) {
          // If request is successful, go to Home page
          this.props.getItems(
            this.props.selectedCategory,
            PaginationConfig.DEFAULT_OFFSET,
            PaginationConfig.ITEMS_PER_PAGE,
          );
          this.props.history.push(`/categories/${this.props.selectedCategory}/items/1`);
        } else {
          this.setState({
            alerts: res.errors.description,
          });
        }
      });
  }

  render() {
    return (
      <div>
        <p name='name'>
          {this.props.item.name}
        </p>
        <p name='price'>
          Price:
          {this.props.item.price}
        </p>
        <p name='description'>
          Description:
          {this.props.item.description}
        </p>
        { this.props.isAuthorized
        && (
        <div>
          <Link
            name='updateLink'
            to={`/items/${this.props.item.id}/update`}
          >
          Update
          </Link>
          <Button
            name='removeBtn'
            onClick={this.deleteItem}
          >
          Remove
          </Button>
        </div>
        )
        }
        {
          this.state.alerts
          && (
          <Alert variant='danger'>
            {this.state.alerts}
          </Alert>
          )
        }
      </div>
    );
  }
}

function mapStateToProps({ items, categories, users }, { match }) {
  const itemId = Number(match.params.id);
  const selectedItem = Object.values(items.items)
    .find(item => item.id === itemId);
  const isAuthorized = Number(users.userId) === selectedItem.user_id;
  return {
    item: selectedItem,
    selectedCategory: categories.currentCategory,
    isAuthorized,
  };
}

const mapDispatchToProps = {
  getItems,
  removeItem,
};

export default connect(mapStateToProps, mapDispatchToProps)(ItemDetail);
