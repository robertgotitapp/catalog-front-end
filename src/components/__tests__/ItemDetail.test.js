import React from 'react';
import { shallow } from 'enzyme';
import Alert from 'react-bootstrap/Alert';
import { ItemDetail } from '../ItemDetail';

describe('components/ItemDetail', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      item: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 11,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      selectedCategory: {
        created: '2019-06-17T13:46:18+00:00',
        description: 'Laptop',
        id: 1,
        name: 'Laptop',
        updated: '2019-06-17T13:46:18+00:00',
      },
      currentPage: 1,
      getItems: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      removeItem: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      selectItemPage: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      isAuthorized: true,
    };
  });

  const clickDeleteBtn = () => {
    wrapper.instance().deleteItem({
      preventDefault: jest.fn(),
    });
  };

  const setup = () => {
    wrapper = shallow(<ItemDetail {...props} />);
  };

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when user is not authorized to change or user is not logged in', () => {
    props = {
      item: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 11,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      selectedCategory: {
        created: '2019-06-17T13:46:18+00:00',
        description: 'Laptop',
        id: 1,
        name: 'Laptop',
        updated: '2019-06-17T13:46:18+00:00',
      },
      currentPage: 1,
      getItems: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      removeItem: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      selectItemPage: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      isAuthorized: false,
    };
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should set toHome state to true when delete item successfully', async () => {
    setup();
    clickDeleteBtn();
    await Promise.resolve();
    expect(wrapper.state().toHome).toEqual(true);
  });

  it('should trigger alert when delete item request is failed', async () => {
    props = {
      item: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 11,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      selectedCategory: {
        created: '2019-06-17T13:46:18+00:00',
        description: 'Laptop',
        id: 1,
        name: 'Laptop',
        updated: '2019-06-17T13:46:18+00:00',
      },
      currentPage: 1,
      getItems: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
      removeItem: jest.fn(() => Promise.resolve({
        statusCode: 0,
        errorPromise: Promise.resolve({
          description: 'Failed to delete item',
        }),
      })),
      selectItemPage: jest.fn(() => Promise.resolve({
        statusCode: 1,
      })),
    };
    setup();
    clickDeleteBtn();
    await Promise.resolve();
    expect(wrapper.find(Alert)).toHaveLength(1);
  });
});
