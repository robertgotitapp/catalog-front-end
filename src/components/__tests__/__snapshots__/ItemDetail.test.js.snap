// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddItem should render correctly 1`] = `
<div>
  <p
    name="name"
  >
    Item 4
  </p>
  <p
    name="price"
  >
    Price:
    19
  </p>
  <p
    name="description"
  >
    Description:
    Item 4
  </p>
</div>
`;
